File: Programmers/Lv2/게임맵최단거리/Solution.java
package Lv2.게임맵최단거리;

import java.util.LinkedList;
import java.util.Queue;

class Solution {
    private static final int[] dx = {0, 0, 1, -1};
    private static final int[] dy = {1, -1, 0, 0};

    private static class Dot {
        public final int x;
        public final int y;

        public Dot(int y, int x) {
            this.x = x;
            this.y = y;
        }
    }

    private int BFS(int[][] maps) {
        int row = maps.length - 1;
        int col = maps[0].length - 1;

        Queue<Dot> q = new LinkedList<>();
        q.add(new Dot(0, 0));

        while (!q.isEmpty()) {
            Dot dot = q.poll();
            int cost = maps[dot.y][dot.x];

            for (int i = 0; i < 4; i++) {
                int ny = dot.y + dy[i];
                int nx = dot.x + dx[i];

                if (nx < 0 || nx > col || ny < 0 || ny > row) continue;
                else if (maps[ny][nx] == 0) continue;

                if (maps[ny][nx] == 1 || maps[ny][nx] > cost + 1) {
                    maps[ny][nx] = cost + 1;
                    q.add(new Dot(ny, nx));
                }
            }
        }

        return maps[row][col];
    }

    public int solution(int[][] maps) {
        int answer = BFS(maps);
        return answer > 1 ? answer : -1;
    }
}

File: Programmers/Lv2/게임맵최단거리/solution.py
from collections import deque


def solution(maps):
    row = len(maps)
    col = len(maps[0])

    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]

    dq = deque([(0, 0)])

    while dq:
        y, x = dq.popleft()
        cost = maps[y][x]

        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]

            if nx < 0 or nx >= col or ny < 0 or ny >= row:
                continue
            elif maps[ny][nx] == 0:
                continue

            if maps[ny][nx] == 1 or maps[ny][nx] > cost + 1:
                maps[ny][nx] = cost + 1
                dq.append([ny, nx])


    answer = maps[row - 1][col - 1]

    return answer if answer > 1 else -1
